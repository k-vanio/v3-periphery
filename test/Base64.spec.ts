import { ethers } from 'hardhat'
import { expect } from './shared/expect'
import { Base64Test } from '../typechain'
import { randomBytes } from 'crypto'

function stringToHex(str: string): string {
  return `0x${Buffer.from(str, 'utf8').toString('hex')}`
}

function base64Encode(str: string): string {
  return Buffer.from(str, 'utf8').toString('base64')
}

describe('Base64', () => {
  let base64: Base64Test
  before('deploy test contract', async () => {
    base64 = (await (await ethers.getContractFactory('Base64Test')).deploy()) as Base64Test
  })

  describe('#encode', () => {
    it('is correct for empty bytes', async () => {
      expect(await base64.encode(stringToHex(''))).to.eq('')
    })

    for (const example of [
      'test string',
      'this is a test',
      'alphabet soup',
      'aLpHaBeT',
      'includes\nnewlines',
      '<some html>',
      'ðŸ˜€',
      'f',
      'fo',
      'foo',
      'foob',
      'fooba',
      'foobar',
    ]) {
      it(`works for "${example}"`, async () => {
        expect(await base64.encode(stringToHex(example))).to.eq(base64Encode(example))
      })
    }

    it('tiny fuzzing', async () => {
      const inputs = []
      for (let i = 0; i < 100; i++) {
        inputs.push(randomBytes(Math.random() * 100))
      }

      const promises = inputs.map((input) => {
        return base64.encode(`0x${input.toString('hex')}`)
      })

      const results = await Promise.all(promises)

      for (let i = 0; i < inputs.length; i++) {
        expect(inputs[i].toString('base64')).to.eq(results[i])
      }
    }).timeout(300_000)
  })
})
