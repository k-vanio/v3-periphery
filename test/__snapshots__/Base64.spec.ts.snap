// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base64 #encode gas cost of encode(<some html>) 1`] = `8361`;

exports[`Base64 #encode gas cost of encode(aLpHaBeT) 1`] = `6256`;

exports[`Base64 #encode gas cost of encode(alphabet soup) 1`] = `10038`;

exports[`Base64 #encode gas cost of encode(f) 1`] = `1631`;

exports[`Base64 #encode gas cost of encode(fo) 1`] = `2056`;

exports[`Base64 #encode gas cost of encode(foo) 1`] = `2395`;

exports[`Base64 #encode gas cost of encode(foob) 1`] = `3732`;

exports[`Base64 #encode gas cost of encode(fooba) 1`] = `4154`;

exports[`Base64 #encode gas cost of encode(foobar) 1`] = `4497`;

exports[`Base64 #encode gas cost of encode(includes
newlines) 1`] = `12564`;

exports[`Base64 #encode gas cost of encode(test string) 1`] = `8358`;

exports[`Base64 #encode gas cost of encode(this is a test) 1`] = `10461`;

exports[`Base64 #encode gas cost of encode(this is a very long string that should cost a lot of gas to encode :)) 1`] = `48770`;

exports[`Base64 #encode gas cost of encode(ðŸ˜€) 1`] = `3732`;
